File: 04-Linked-List/Add-Two-Numbers-pblmNo-02/solution.py
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    def addTwoNumbers(self, l1, l2):
        dummy = ListNode(0)
        curr = dummy
        carry = 0

        while l1 or l2 or carry:
            total = carry

            if l1:
                total += l1.val
                l1 = l1.next

            if l2:
                total += l2.val
                l2 = l2.next

            carry = total // 10
            curr.next = ListNode(total % 10)
            curr = curr.next

        return dummy.next


# ---------- Helper Functions ----------

def createLinkedList(arr):
    dummy = ListNode()
    curr = dummy

    for num in arr:
        curr.next = ListNode(num)
        curr = curr.next

    return dummy.next


def printLinkedList(head):
    while head:
        print(head.val, end=" -> " if head.next else "")
        head = head.next
    print()


# ---------- Driver Code ----------

l1 = createLinkedList([2, 4, 3])
l2 = createLinkedList([5, 6, 4])

solution = Solution()
result = solution.addTwoNumbers(l1, l2)

print("Result:")
printLinkedList(result)

File: 04-Linked-List/Add-Two-Numbers-pblmNo-02/solution.js
class ListNode {
    constructor(val = 0, next = null) {
        this.val = val;
        this.next = next;
    }
}

class Solution {
    addTwoNumbers(l1, l2) {
        const dummy = new ListNode(0);
        let curr = dummy;
        let carry = 0;

        while (l1 || l2 || carry) {
            let total = carry;

            if (l1) {
                total += l1.val;
                l1 = l1.next;
            }

            if (l2) {
                total += l2.val;
                l2 = l2.next;
            }

            carry = Math.floor(total / 10);

            curr.next = new ListNode(total % 10);
            curr = curr.next;
        }

        return dummy.next;
    }
}

// ---------- Helper Functions ----------

function createLinkedList(arr) {
    const dummy = new ListNode();
    let curr = dummy;

    for (const num of arr) {
        curr.next = new ListNode(num);
        curr = curr.next;
    }

    return dummy.next;
}

function printLinkedList(head) {
    const result = [];

    while (head) {
        result.push(head.val);
        head = head.next;
    }

    console.log(result.join(" -> "));
}

// ---------- Driver Code ----------

const l1 = createLinkedList([2, 4, 3]);
const l2 = createLinkedList([5, 6, 4]);

const solution = new Solution();
const result = solution.addTwoNumbers(l1, l2);

console.log("Result:");
printLinkedList(result);